readme.md
1. npm install
2. npm run migrate-kovan

To check unit tests coverage please run:
- truffle run coverage

To run unit tests you need to do 2 commands one by one:
- truffle develop
- test

changelog.md
## 04.02.2021

### StakeMaster

#### Changed
- StakingPoolCreated event add all pools arguments
- BurnPercentUpdated event add devider


### StakingPool

#### Added 
- allStakedAmount
- allPaidReward
- allRewardDebt
- poolTokenAmount
- participants
- withdrawPoolRemainder method return not distributed tokens for owner
- allow to create pool with the same token (reward, stake)


#### Updated
- stakes,rewardDebts mapping was removed, add mapping with userInfo with struct UserInfo (amount, rewardDebt)
- reward mul changed from 1e12 to 1e18

#### Removed
- Update finish block
- Update Pool token amount
- Withdraw all reward from pool by owner

test/StakeMasterTest.js

const { assert } = require("chai");
const { expectEvent, expectRevert } = require("openzeppelin-test-helpers");

const ERC20Basic = artifacts.require("ERC20Basic");
const StakeMaster = arti.require("StakeMaster");

contract("StakeMaster", function (accounts) {
  const totalSupply = "10000000000000000000000"; // 10000
  const poolTokenSupply = "10000000000000000000000"; // 10000
  const feeAmount = "10000000000000000000"; // 10
  const newFeeAmount = "15000000000000000000"; // 15
  const burnPercent = "1";
  const poolDurationInBlocks = 2000;

  let feeToken;
  let stakingToken;
  let poolToken;
  let stakeMaster;

  beforeEach(async function () {
    feeToken = await ERC20Basic.new(totalSupply);
    stakingToken = await ERC20Basic.new(totalSupply, { from: accounts[2] });
    poolToken = await ERC20Basic.new(totalSupply);

    stakeMaster = await StakeMaster.new(
      feeToken.address,
      accounts[1],
      feeAmount,
      burnPercent
    );
  });

  describe("createStakingPool", function () {
    it("Staking pool should be created successfully", async function () {
      await feeToken.approve(stakeMaster.address, totalSupply);
      await poolToken.approve(stakeMaster.address, poolTokenSupply);

      let startBlock = (await web3.eth.getBlock("latest")).number;
      let finishBlock = startBlock + poolDurationInBlocks;

      let tx = await stakeMaster.createStakingPool(
        stakingToken.address,
        poolToken.address,
        startBlock,
        finishBlock,
        poolTokenSupply
      );

      expectEvent.inLogs(tx.logs, "StakingPoolCreated", {
        owner: accounts[0],
          pool: tx.logs[tx.logs.length - 1].args.pool,
          stakingToken: stakingToken.address,
          poolToken: poolToken.address,
          startBlock: startBlock,
          finishBlock: finishBlock,
          poolTokenAmount: poolTokenSupply
      });
    });

    it("Transaction should fail due to caller doesnt give allowance of fee token to Staking Master contract", async function () {
      await poolToken.approve(stakeMaster.address, poolTokenSupply);

      let startBlock = (await web3.eth.getBlock("latest")).number;
      let finishBlock = startBlock + poolDurationInBlocks;

      await expectRevert(
        stakeMaster.createStakingPool(
          stakingToken.address,
          poolToken.address,
          startBlock,
          finishBlock,
          poolTokenSupply
        ),
        "revert"
      );
    });

    it("Transaction should fail due to caller doesnt give allowance of staking token to Staking Master contract", async function () {
      await feeToken.approve(stakeMaster.address, totalSupply);

      let startBlock = (await web3.eth.getBlock("latest")).number;
      let finishBlock = startBlock + poolDurationInBlocks;

      await expectRevert(
        stakeMaster.createStakingPool(
          stakingToken.address,
          poolToken.address,
          startBlock,
          finishBlock,
          poolTokenSupply
        ),
        "revert"
      );
    });

    it("Transaction should fail due to caller doesnt have enough tokens for fee payment for staking pool creation", async function () {
      await feeToken.approve(stakeMaster.address, totalSupply);
      await poolToken.approve(stakeMaster.address, poolTokenSupply);

      let startBlock = (await web3.eth.getBlock("latest")).number;
      let finishBlock = startBlock + poolDurationInBlocks;

      await feeToken.transfer(accounts[3], totalSupply);

      await expectRevert(
        stakeMaster.createStakingPool(
          stakingToken.address,
          poolToken.address,
          startBlock,
          finishBlock,
          poolTokenSupply
        ),
        "revert"
      );
    });

    it("Transaction should fail due to caller doesnt have enough pool tokens for staking pool creation", async function () {
      await feeToken.approve(stakeMaster.address, totalSupply);
      await poolToken.approve(stakeMaster.address, poolTokenSupply);

      let startBlock = (await web3.eth.getBlock("latest")).number;
      let finishBlock = startBlock + poolDurationInBlocks;

      await poolToken.transfer(accounts[3], totalSupply);

      await expectRevert(
        stakeMaster.createStakingPool(
          stakingToken.address,
          poolToken.address,
          startBlock,
          finishBlock,
          poolTokenSupply
        ),
        "revert"
      );
    });

    it("Transaction should fail due to startBlock of staking pool must be less than finishBlock", async function () {
      await feeToken.approve(stakeMaster.address, totalSupply);
      await poolToken.approve(stakeMaster.address, poolTokenSupply);

      let startBlock = (await web3.eth.getBlock("latest")).number;
      let finishBlock = startBlock;

      await expectRevert(
        stakeMaster.createStakingPool(
          stakingToken.address,
          poolToken.address,
          startBlock,
          finishBlock,
          poolTokenSupply
        ),
        "start block must be less than finish"
      );
    });

    it("Transaction should fail due to startBlock of staking pool must be less than finishBlock", async function () {
      await feeToken.approve(stakeMaster.address, totalSupply);
      await poolToken.approve(stakeMaster.address, poolTokenSupply);

      let startBlock = (await web3.eth.getBlock("latest")).number - 1;
      let finishBlock = (await web3.eth.getBlock("latest")).number;

      await expectRevert(
        stakeMaster.createStakingPool(
          stakingToken.address,
          poolToken.address,
          startBlock,
          finishBlock,
          poolTokenSupply
        ),
        "finish block must be more than current bloc"
      );
    });
  });

  describe("setFeeAmount", function () {
    it("New fee amount should be setted successfully", async function () {
      let tx = await stakeMaster.setFeeAmount(newFeeAmount);

      expectEvent.inLogs(tx.logs, "FeeAmountUpdated", { newFeeAmount });

      assert.equal(newFeeAmount, await stakeMaster.feeAmount());
    });

    it("Transaction should fail due to caller is not an owner", async function () {
      await expectRevert(
        stakeMaster.setFeeAmount(newFeeAmount, { from: accounts[1] }),
        "revert"
      );
    });
  });

  describe("setFeeToken", function () {
    it("New fee token address should be setted successfully", async function () {
      let newFeeToken = (await ERC20Basic.new(totalSupply)).address;
      let tx = await stakeMaster.setFeeToken(newFeeToken);

      expectEvent.inLogs(tx.logs, "TokenFeeUpdated", { newFeeToken });

      assert.equal(newFeeToken, await stakeMaster.feeToken());
    });

    it("Transaction should fail due to new fee token address is not a contract address", async function () {
      await expectRevert(
        stakeMaster.setFeeToken(accounts[3], { from: accounts[1] }),
        "revert"
      );
    });

    it("Transaction should fail due to caller is not an owner", async function () {
      let newFeeToken = (await ERC20Basic.new(totalSupply)).address;

      await expectRevert(
        stakeMaster.setFeeToken(newFeeToken, { from: accounts[1] }),
        "revert"
      );
    });
  });

  describe("setFeeWallet", function () {
    it("New fee wallet should be setted successfully", async function () {
      let tx = await stakeMaster.setFeeWallet(accounts[5]);

      expectEvent.inLogs(tx.logs, "FeeWalletUpdated", {
        newFeeWallet: accounts[5],
      });

      assert.equal(accounts[5], await stakeMaster.feeWallet());
    });

    it("Transaction should fail due to caller is not an owner", async function () {
      await expectRevert(
        stakeMaster.setFeeWallet(accounts[5], { from: accounts[1] }),
        "revert"
      );
    });
  });

  describe("setBurnPercent", function () {
    it("New fee wallet should be setted successfully", async function () {
      let tx = await stakeMaster.setBurnPercent("2", "100");

      expectEvent.inLogs(tx.logs, "BurnPercentUpdated", {
        newBurnPercent: "2",
      });

      assert.equal("2", await stakeMaster.burnPercent());
    });

    it("Transaction should fail due to caller is not an owner", async function () {
      await expectRevert(
        stakeMaster.setBurnPercent("2", "100", { from: accounts[1] }),
        "revert"
      );
    });
  });
});


test/TokenVesting.test.js
const {
  BN,
  constants,
  expectEvent,
  ether,
  expectRevert,
  time,
} = require("@openzeppelin/test-helpers");
const { ZERO_ADDRESS } = constants;
const { expect } = require("chai");
const timeTravel = time.increaseTo;

const TokenVesting = artifacts.require("TokenVesting");
const ERC20Mock = artifacts.require("ERC20Mock");

contract.only("TokenVesting", function (accounts) {
  const [owner, beneficiary, other] = accounts;

  const ownerBal = ether("12000000000000000000");
  let start;
  const interval = 240; // 4 minutes
  const duration = interval * 4; // 16 minutes
  const balance = ether("120"); // 120 tokens

  beforeEach(async function () {
    const blockNumber = await web3.eth.getBlockNumber();
    const block = await web3.eth.getBlock(blockNumber);
    start = block.timestamp;
    this.token = await ERC20Mock.new(owner, ownerBal);
    this.tokenVesting = await TokenVesting.new(this.token.address);
    this.token.approve(this.tokenVesting.address, ownerBal, { from: owner });
  });

  const mineBlock = async function (_token, spender) {
    await _token.burn(owner, "0", { from: owner });
  };

  describe("create vesting", function () {
    beforeEach("createVesting", async function () {
      const allowance = await this.token.allowance(
        owner,
        this.tokenVesting.address
      );
      // console.log('allowance', allowance.toString())
      // console.log('balance', balance.toString())
      await this.tokenVesting.createVesting(
        owner,
        beneficiary,
        start,
        interval,
        duration,
        balance
      );
    });

    describe("vesting should have correct data", function () {
      it("getVesting", async function () {
        const vesting = await this.tokenVesting.getVesting(beneficiary);
        expect(vesting[0]).to.be.bignumber.equal(start.toString());
        expect(vesting[1]).to.be.bignumber.equal(interval.toString());
        expect(vesting[2]).to.be.bignumber.equal(duration.toString());
        expect(vesting[3]).to.be.bignumber.equal(balance.toString());
        expect(vesting[4]).to.be.bignumber.equal("0");
      });
    });

    describe("reverts", function () {
      it("when interval > duration", async function () {
        await expectRevert(
          this.tokenVesting.createVesting(
            owner,
            beneficiary,
            start,
            interval + 1,
            interval,
            balance
          ),
          "TokenVesting #createVesting: interval cannot be bigger than duration"
        );
      });

      it("when current balance > 0", async function () {
        await expectRevert(
          this.tokenVesting.createVesting(
            owner,
            beneficiary,
            start,
            interval,
            duration,
            balance
          ),
          "TokenVesting #createVesting: vesting for beneficiary already created"
        );
      });
    });

    const checkVesting = function (intervalsVested, alreadyReleased) {
      describe(`should vested ${intervalsVested} intervals`, async function () {
        it("vestedAmount", async function () {
          const vestedAmount = await this.tokenVesting.vestedAmount(
            beneficiary
          );
          const intervalsVestedBN = new BN(intervalsVested.toString());
          const durationBN = new BN(duration.toString());
          const intervalBN = new BN(interval.toString());
          const alreadyReleasedBN = new BN(alreadyReleased.toString());
          const expectedVestedAmount = balance
            .mul(intervalsVestedBN)
            .div(durationBN.div(intervalBN));
          expect(vestedAmount).to.be.bignumber.equal(expectedVestedAmount);
        });

        it("releasableAmount", async function () {
          const releasableAmount = await this.tokenVesting.releasableAmount(
            beneficiary
          );
          const intervalsVestedBN = new BN(intervalsVested.toString());
          const durationBN = new BN(duration.toString());
          const intervalBN = new BN(interval.toString());
          const alreadyReleasedBN = new BN(alreadyReleased.toString());
          const expectedReleasableAmount = balance
            .mul(intervalsVestedBN.sub(alreadyReleasedBN))
            .div(durationBN.div(intervalBN));
          expect(releasableAmount).to.be.bignumber.equal(
            expectedReleasableAmount
          );
        });
      });
    };

    const releaseAndCheckBalances = function (
      intervalsVested,
      alreadyReleased
    ) {
      describe(`should release token for ${intervalsVested} intervals`, function () {
        it("release", async function () {
          const releasableAmount = await this.tokenVesting.releasableAmount(
            beneficiary
          );

          const openBalBeneficiary = await this.token.balanceOf(beneficiary);
          const openBalVestingContract = await this.token.balanceOf(
            this.tokenVesting.address
          );
          const openBalVesting = new BN(
            (await this.tokenVesting.getVesting(beneficiary))[3].toString()
          );

          await this.tokenVesting.release(beneficiary);

          const closeBalBeneficiary = await this.token.balanceOf(beneficiary);
          const closeBalVestingContract = await this.token.balanceOf(
            this.tokenVesting.address
          );
          const closeBalVesting = new BN(
            (await this.tokenVesting.getVesting(beneficiary))[3].toString()
          );

          expect(closeBalBeneficiary).to.be.bignumber.equal(
            openBalBeneficiary.add(releasableAmount)
          );
          expect(closeBalVestingContract).to.be.bignumber.equal(
            openBalVestingContract.sub(releasableAmount)
          );
          expect(closeBalVesting).to.be.bignumber.equal(
            openBalVesting.sub(releasableAmount)
          );
        });
      });
    };

    context("time travel 1 interval", function () {
      beforeEach("time travel", async function () {
        await mineBlock(this.token, this.tokenVesting.address);
        await timeTravel(start + interval);
        await mineBlock(this.token, this.tokenVesting.address);
      });

      checkVesting(1, 0);
      releaseAndCheckBalances(1, 0);
    });

    context("time travel 2 intervals minus 10 second", function () {
      beforeEach("time travel", async function () {
        await mineBlock(this.token, this.tokenVesting.address);
        await timeTravel(start + interval * 2 - 10);
        await mineBlock(this.token, this.tokenVesting.address);
      });
      context("when released for interval 1", function () {
        beforeEach("release for interval 1", async function () {
          await this.tokenVesting.release(beneficiary);
        });
        checkVesting(1, 1);
      });
    });

    context("time travel 2 intervals", function () {
      beforeEach("time travel", async function () {
        await mineBlock(this.token, this.tokenVesting.address);
        await timeTravel(start + interval * 2 + 5);
        await mineBlock(this.token, this.tokenVesting.address);
      });
      checkVesting(2, 0);
      releaseAndCheckBalances(2, 0);
    });

    context("time travel 2 intervals and release after each", function () {
      beforeEach("time travel", async function () {
        await mineBlock(this.token, this.tokenVesting.address);
        await timeTravel(start + interval + 5);
        await this.tokenVesting.release(beneficiary);
        await timeTravel(start + interval * 2 + 5);
        await mineBlock(this.token, this.tokenVesting.address);
      });
      checkVesting(2, 1);
      releaseAndCheckBalances(2, 1);
    });

    context("time travel 3 intervals and release", function () {
      beforeEach("time travel", async function () {
        await mineBlock(this.token, this.tokenVesting.address);
        await timeTravel(start + interval * 3 + 5);
        await mineBlock(this.token, this.tokenVesting.address);
      });
      checkVesting(3, 0);
      releaseAndCheckBalances(3, 0);
    });

    context("time travel 4 intervals, postpone and release", function () {
      let newStart, wrongNewStart;

      beforeEach("time travel", async function () {
        newStart = start + interval * 3;
        wrongNewStart = start - 1;
        await mineBlock(this.token, this.tokenVesting.address);
        await timeTravel(start + interval * 4 + 5);
        await mineBlock(this.token, this.tokenVesting.address);
      });
      checkVesting(4, 0);

      context("postpone vesting", function () {
        describe("reverts", function () {
          it("when balance is 0 (vesting not extists)", async function () {
            await expectRevert(
              this.tokenVesting.postponeVesting(newStart, { from: owner }),
              "TokenVesting #postponeVesting: vesting for beneficiary does not exist"
            );
          });

          it("new start is before old start", async function () {
            await expectRevert(
              this.tokenVesting.postponeVesting(start, { from: beneficiary }),
              "TokenVesting #postponeVesting: new start date cannot be earlier than original start date"
            );

            await expectRevert(
              this.tokenVesting.postponeVesting(wrongNewStart, {
                from: beneficiary,
              }),
              "TokenVesting #postponeVesting: new start date cannot be earlier than original start date"
            );
          });
        });

        it("postponeVesting", async function () {
          await this.tokenVesting.postponeVesting(newStart, {
            from: beneficiary,
          });
          const vesting = await this.tokenVesting.getVesting(beneficiary);

          expect(vesting[0]).to.be.bignumber.equal(newStart.toString());
          expect(vesting[1]).to.be.bignumber.equal(interval.toString());
          expect(vesting[2]).to.be.bignumber.equal(duration.toString());
          expect(vesting[3]).to.be.bignumber.equal(balance.toString());
          expect(vesting[4]).to.be.bignumber.equal("0");
        });

        context("postpone 3 intervals", function () {
          beforeEach(async function () {
            await this.tokenVesting.postponeVesting(newStart, {
              from: beneficiary,
            });
          });
          checkVesting(1, 0);
          releaseAndCheckBalances(1, 0);
        });
      });
    });

    context("time travel 4 intervals and release after each", function () {
      beforeEach("time travel", async function () {
        await mineBlock(this.token, this.tokenVesting.address);
        await timeTravel(start + interval + 5);
        await this.tokenVesting.release(beneficiary);
        await timeTravel(start + interval * 2 + 5);
        await this.tokenVesting.release(beneficiary);
        await timeTravel(start + interval * 3 + 5);
        await this.tokenVesting.release(beneficiary);
        await timeTravel(start + interval * 4 + 5);
        await mineBlock(this.token, this.tokenVesting.address);
      });
      checkVesting(4, 3);
      releaseAndCheckBalances(4, 3);
    });

    context("time travel 4 intervals and release", function () {
      beforeEach("time travel", async function () {
        await mineBlock(this.token, this.tokenVesting.address);
        await timeTravel(start + interval * 4 + 5);
        await mineBlock(this.token, this.tokenVesting.address);
      });
      checkVesting(4, 0);
      releaseAndCheckBalances(4, 0);
    });
  });
});


script.js
const axios = require("axios");
const fs = require("fs");

const tokenAddress = "0xE90334c19c798C3A18d81b8cd16594247D5B19dd";
const apiEndpoint = "https://api.bscscan.com/api";
const apiKey = "";

const getTransactions = async () => {
  try {
    let result = [];
    let page = 1;

    while (true) {
      const response = await axios.get(apiEndpoint, {
        params: {
          module: "account",
          action: "tokentx",
          contractaddress: tokenAddress,
          page: page,
          offset: 100,
          sort: "desc",
          apikey: apiKey,
        },
      });

      const { data } = response;

      if (data.status !== "1" || !data.result.length) {
        break;
      }

      result = result.concat(data.result);
      page++;

      await new Promise((resolve) => setTimeout(resolve, 1000));
    }

    return result;
  } catch (error) {
    console.error(error);
  }
};

getTransactions()
  .then((result) => {
    fs.writeFile("transactions.json", JSON.stringify(result), (err) => {
      if (err) {
        console.error(err);
        return;
      }
      console.log("Transactions saved to transactions.json");
    });
  })
  .catch((error) => {
    console.error(error);
  });


truffle-config.js
const HDWalletProvider = require("truffle-hdwallet-provider");


module.exports = {
    networks: {
        development: {
            host: "127.0.0.1",
            port: 8545,
            network_id: "*",
            //TODO: change to IDO contracts
            gas: 8000000,
        },
        kovan: {
            // must be a thunk, otherwise truffle commands may hang in CI
            provider: () => new HDWalletProvider("", "https://kovan.infura.io/v3/46e5f1638bb04dd4abb7f75bfd4f8898"),
            network_id: "42",
            from: "0xb58967989C8e878de4D7e78965e066F26B2d9bF4",
            gasPrice: 35000000000,
            skipDryRun: true,
        },
        testBSC: {
            // must be a thunk, otherwise truffle commands may hang in CI
            provider: () => new HDWalletProvider("", "https://data-seed-prebsc-1-s1.binance.org:8545/"),
            network_id: 97,
            from: "0xb58967989C8e878de4D7e78965e066F26B2d9bF4",
            skipDryRun: true
        },
        BSC: {
            // must be a thunk, otherwise truffle commands may hang in CI
            provider: () => new HDWalletProvider([""], "https://bsc-dataseed.binance.org/"),
            network_id: "56",
            from: "",
            skipDryRun: true,
            //gas: 8000000,
            //gasPrice: 11000000000,// 11 gwei
        },
        AvaxTestnet: {
            // must be a thunk, otherwise truffle commands may hang in CI
            provider: () => new HDWalletProvider("", "https://api.avax-test.network/ext/bc/C/rpc"),
            network_id: "43113",
            from: "0xb58967989C8e878de4D7e78965e066F26B2d9bF4",
            skipDryRun: true,
            gasPrice: 22500000000,// 22,5 gwei
        },
        ETH: {
            // must be a thunk, otherwise truffle commands may hang in CI
            provider: () => new HDWalletProvider([""], "https://mainnet.infura.io/v3/46e5f1638bb04dd4abb7f75bfd4f8898"),
            network_id: "1",
            from: "",
            skipDryRun: true,
            gasPrice: 22000000000,// 25 gwei
        },
        Fantom: {
            //https://explorer.testnet.fantom.network/
            // must be a thunk, otherwise truffle commands may hang in CI
            provider: () => new HDWalletProvider("", "https://rpc.fantom.network/"),
            network_id: "250",
            from: "",
            skipDryRun: true,
            gasPrice: 52000000000,// 52 gwei
        },
        testFantom: {
            //https://explorer.testnet.fantom.network/
            // must be a thunk, otherwise truffle commands may hang in CI
            provider: () => new HDWalletProvider("", "https://rpc.testnet.fantom.network/"),
            network_id: "4002",
            from: "",
            skipDryRun: true,
        },
    },
    plugins: ['truffle-plugin-verify'],
    api_keys: {
        bscscan: '',
      },
    mocha: {
        useColors: true,
    },
    compilers: {
        solc: {
            version: "0.6.12",
            optimizer: {
                enabled: true,
                runs: 200,
            },
        },
    },
    plugins: ["truffle-plugin-verify", "solidity-coverage", "truffle-contract-size"],
    api_keys: {
        etherscan: "",
        bscscan: "QTU498KHG8T1HHQQKDFKA82D2DUZ54R3U3",
        ftmscan: '',
    },
};


migrations/1_deploy_IDO.js
const ERC20Basic = artifacts.require("ERC20Basic");
const FeeToken = artifacts.require("FeeToken");
const StakeMaster = artifacts.require("StakeMaster");
const StakingPool = artifacts.require("StakingPool");
const IDOMaster = artifacts.require("IDOMaster");
const IDOCreator = artifacts.require("IDOCreator");
const IDOPool = artifacts.require("IDOPool");
const TierSystem = artifacts.require("TierSystem");
const { toWei, fromWei, toBN } = web3.utils;

let idoMasterContract;

const totalSupply = toWei("10000");
const poolTokenSupply = toWei("1");
const stakeAmount = toWei("10");
const feeAmount = toWei("0");
const poolDurationInSecunds = 900;




//TestBSC
const feeWallet = "0xb58967989C8e878de4D7e78965e066F26B2d9bF4";
const feeToken = "0xce01c35b316ccdf43eb3a5f73aa597d519637a28";
const burnPercent = "0";



module.exports = async (deployer, network, accounts) => {
    
    await deployer.deploy(TierSystem, toWei("100"), "100", toWei("10"), "100", toWei("100"), "100",toWei("100"), "100", toWei("100"), "100");
    let tierSystemContract = await TierSystem.deployed();
    console.log("TierSystem address ====> " + tierSystemContract.address);

    await deployer.deploy(IDOMaster, feeToken, feeWallet, feeAmount, burnPercent);
    let idoMasterContract = await IDOMaster.deployed();
    console.log("IDO master address ====> " + idoMasterContract.address);

    await deployer.deploy(IDOCreator, idoMasterContract.address, tierSystemContract.address);
    let idoCreatorContract = await IDOCreator.deployed();
    console.log("IDO creator address ====> " + idoCreatorContract.address);

    await idoMasterContract.setCreatorProxy(idoCreatorContract.address);
    console.log("set setCreatorProxy ====> " + idoCreatorContract.address);
};
